### PR TITLE
Allow children to have boundary conditions

### PR DESCRIPTION
**Pull request by grmnptr**
Taking over #3470 which took over #3094

### CODE REVIEW

File: src/mesh/boundary_info.C
Comment (roystgnr) [inline src/mesh/boundary_info.C:1252]:
```suggestion
  // need to store BCs too. This case is covered with the _children_on_boundary
```

File: src/mesh/boundary_info.C
Comment (roystgnr) [inline src/mesh/boundary_info.C:1250]:
```suggestion
  // In most cases only level-0 elements store BCs.
```

File: src/mesh/boundary_info.C
Comment (roystgnr) [inline src/mesh/boundary_info.C:1280]: This is redundant with the if test above.  The redundant for block above could be eliminated, too, by doing `while (searched_elem)` so you get both the element and its ancestors with the same entry.

File: src/mesh/boundary_info.C
Comment (roystgnr) [inline src/mesh/boundary_info.C:1778]:
```suggestion
  // query the parent boundary ID on a child.
```

File: include/mesh/boundary_info.h
Comment (roystgnr) [inline include/mesh/boundary_info.h:572]: This is somewhat vague and confusing - I can't tell from the comment whether `elem` is expected to be the parent or each child, from the function name I'd expect the child, from the function implementation I'd expect the parent (except that non-parents are filtered out rather than being errors?).

Comment (roystgnr) [reply include/mesh/boundary_info.h:572]: Should we change the name to `transfer_boundary_ids_from_children`, change the parameter name to `parent`?

File: include/mesh/boundary_info.h
Comment (roystgnr) [inline include/mesh/boundary_info.h:569]: Surely "two" should be dimension-dependent?  If we're in 1D, for example, and there's only 1 child on a side to begin with, then we still definitely want to copy that id back to the parent!  In 2D "2" is reasonable but then the equivalent criterion in 3D would be "4"; alternatively "2" would also be reasonable in 3D but in that case the equivalent criterion in 2D would be "1".

File: src/mesh/boundary_info.C
Comment (roystgnr) [inline src/mesh/boundary_info.C:1890]: Aren't these four lines just exactly what `++boundary_counts[pr.second.second]` would do?

**Review by roystgnr — CHANGES_REQUESTED** [review]

**Comment by roystgnr** [comment]
This is getting really close.  We should probably turn on the distributed dbg checks right before we merge,rather than waiting for the devel->master merge, just to be safe.  Modifying `Packing<Elem>` is not easy, and just because I don't see any major mistakes in the new version here doesn't mean there aren't any.

The one thing I'm still worried about there: we're not doing any communication after transferring BCs from children to parents, and strictly speaking we should be, because otherwise we might fall out of sync on newly coarsened ghost elements getting BCs transferred from newly deleted remote children.  That alone wouldn't be a disaster for most apps, since we generally only query BCs on local elements ... but follow that up with a repartitioning (as we generally do after AMR/C) that turns an out-of-sync ghost element into a local element ... the element's original owner will send it to the new owner, but IIRC we generally *ignore* incoming communications of that sort (except to assert consistency in dbg mode) for data we already expect to have.

File: include/mesh/boundary_info.h
Comment (grmnptr) [reply include/mesh/boundary_info.h:572]: Great idea!

File: include/mesh/boundary_info.h
Comment (grmnptr) [reply include/mesh/boundary_info.h:569]: Oh yes, this came from my 2D mindset. I'll do "if 50%+" of the children on that side have it.

**Comment by grmnptr** [comment]
Got it, thanks for the review! I'll push the changes later today.

File: src/mesh/boundary_info.C
Comment (grmnptr) [reply src/mesh/boundary_info.C:1890]: cool suggestion, thanks!

File: src/mesh/boundary_info.C
Comment (roystgnr) [inline src/mesh/boundary_info.C:1897]: We should never be doing a `remove_side` here, should we?  If the parent doesn't have the side then it's a no-op; if the parent *does* have the side then by definition so does all its children, whether or not the side is set on them *directly*.

File: src/mesh/boundary_info.C
Comment (roystgnr) [inline src/mesh/boundary_info.C:1902]: Isn't this redundant with the `remove()` calls in `MeshRefinement::_coarsen_elements`?

**Review by roystgnr — CHANGES_REQUESTED** [review]
Only two remaining issues I could find.  I wouldn't mind about a trillion more test cases, but this looks to be good enough as a starting point, assuming CI agrees with me later that those 4 lines are redundant.  Pinging @jwpeterson in case he wants to take a look before we merge.

File: src/mesh/boundary_info.C
Comment (grmnptr) [reply src/mesh/boundary_info.C]: I deleted this so coarsening won't delete anything from boundary info anymore.

File: src/mesh/boundary_info.C
Comment (grmnptr) [reply src/mesh/boundary_info.C:1898]: I ended up keeping this and adding a condition for the removal in the coarsen loop. The reason behind this is that we might accidentally remove boundary ids before we transfer them back to the parent. Thus, when we allow boundaries on children we remove them after the transfer process.

File: src/mesh/boundary_info.C
Comment (roystgnr) [inline src/mesh/boundary_info.C:1495]: Okay, I think my only remaining concern is what came to me on Slack yesterday:

> We actually probably ought to throw an error for that remove_side() case, if we're not going to support it yet, the more I think about it.  If a child has id 5 because its parent does, and we try to remove it from the child, then ideally we remove it from the parent and restore it on the other children, less ideally we exit with an error and make the app handle it, but what we do now is silently do nothing, and that's totally wrong.

If someone tries to remove a side that isn't there, our current do-nothing behavior is fine, it's like putting white paint on an already-white wall.  But if someone tries to remove a side that is there via an ancestor, but *isn't a raw side*, our current do-nothing behavior isn't fine.  Put white paint on a blue wall and you don't expect to see blue afterwards, not even if the blue was put on with a large brush and the white is being put on with a smaller brush.

This is a new feature, so it doesn't need to be a *comprehensive* new feature, but we definitely don't want to set up applications for failure if the author mistakenly thinks it's comprehensive or if the author accidentally starts to assume that the missing parts of the feature are missing by design.

Could we modify erase_if to return a std::size_t counting the number of erasures, then (here and in the bcid-specific case below) if we get a 0 from that we'll double-check that `boundary_ids(elem, side, vec)` is empty (or below, that `has_boundary_id(elem, side, id)` is false), then throw `libmesh_not_implemented()` if so?

**Review by roystgnr — CHANGES_REQUESTED** [review]

File: src/mesh/boundary_info.C
Comment (roystgnr) [inline src/mesh/boundary_info.C:1539]: Let's not check _children_on_boundary here?  Trying to remove a child side even when no child sides have been added is still a case we want to scream about.

**Review by roystgnr — APPROVED** [review]
I have that one more trivial issue, but I don't think it's worth a holdup; this looks ready to merge once CI is done.

File: src/mesh/boundary_info.C
Comment (grmnptr) [reply src/mesh/boundary_info.C]: Removed!

**Pull request merged**
